[WebXR Hit Test] XRRay support

fujii opened the pull request:
#### b9d52d2f714fb997187caba6026bfa6ddcf17753
<pre>
[WebXR Hit Test] XRRay support
<a href="https://bugs.webkit.org/show_bug.cgi?id=301255">https://bugs.webkit.org/show_bug.cgi?id=301255</a>

Reviewed by Dan Glastonbury.

Added XRRay support based on the Chromium implementations.
&lt;<a href="https://immersive-web.github.io/hit-test/#xrray-interface">https://immersive-web.github.io/hit-test/#xrray-interface</a>&gt;
&lt;<a href="https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/blink/renderer/modules/xr/xr_ray.cc">https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/blink/renderer/modules/xr/xr_ray.cc</a>&gt;

Tests: imported/w3c/web-platform-tests/webxr/hit-test/xrRay_constructor.https.html
       imported/w3c/web-platform-tests/webxr/hit-test/xrRay_matrix.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/xrRay_constructor.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/xrRay_matrix.https-expected.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/webxr/WebXRRay.cpp:
(WebCore::WebXRRay::create):
(WebCore::WebXRRay::WebXRRay):
(WebCore::WebXRRay::origin):
(WebCore::WebXRRay::direction):
(WebCore::WebXRRay::matrix):
* Source/WebCore/Modules/webxr/WebXRRay.h:

Canonical link: <a href="https://commits.webkit.org/301987@main">https://commits.webkit.org/301987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/754813ca397326648ae0eca89ba922fe1be0ba95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47196 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134625 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129420 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47815 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55722 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97090 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65010 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114233 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37057 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32333 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77998 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108100 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32768 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137109 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41766 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105616 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54718 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105267 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50777 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29227 "Found 1 new test failure: http/tests/misc/ftp-eplf-directory.py (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51787 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19961 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54144 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60231 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53378 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55137 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar commented on Source/WebCore/Modules/webxr/WebXRRay.cpp:
```diff
@@ -40,37 +41,70 @@ WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(WebXRRay);
+        return Exception { ExceptionCode::TypeError };
+    if (direction.w)
+        return Exception { ExceptionCode::TypeError };
+    if (origin.w != 1)
```
There are missing steps from the specs here. "If direction’s [w](https://www.w3.org/TR/webxr-hit-test-1/#dom-xrraydirectioninit-w) is not 0.0, throw a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror) and abort these steps."

fujii replied:
The above `if (direction.w)` checks.

svillar replied:
Ups, overlooked it, sorry for the noise.

svillar commented on Source/WebCore/Modules/webxr/WebXRRay.cpp:
```diff
@@ -40,37 +41,70 @@ WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(WebXRRay);
 const DOMPointReadOnly& WebXRRay::origin()
 {
-    return m_transform->position();
+    return DOMPointReadOnly::fromFloatPoint(m_origin);
```
Origin's w must be 1.0 according to specs. Does the call guarantee that?

fujii replied:
Yes. https://github.com/WebKit/WebKit/blob/0de0214e67147b2e2eef0124c8f732b159fb03b9/Source/WebCore/dom/DOMPointReadOnly.h#L54

svillar reviewed (changes requested):
Looking good, left a few comments

djg reviewed (approved):
:shipit:

fujii committed:
[WebXR Hit Test] XRRay support
https://bugs.webkit.org/show_bug.cgi?id=301255

Reviewed by Dan Glastonbury.

Added XRRay support based on the Chromium implementations.
<https://immersive-web.github.io/hit-test/#xrray-interface>
<https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/blink/renderer/modules/xr/xr_ray.cc>

Tests: imported/w3c/web-platform-tests/webxr/hit-test/xrRay_constructor.https.html
       imported/w3c/web-platform-tests/webxr/hit-test/xrRay_matrix.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/xrRay_constructor.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/xrRay_matrix.https-expected.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/webxr/WebXRRay.cpp:
(WebCore::WebXRRay::create):
(WebCore::WebXRRay::WebXRRay):
(WebCore::WebXRRay::origin):
(WebCore::WebXRRay::direction):
(WebCore::WebXRRay::matrix):
* Source/WebCore/Modules/webxr/WebXRRay.h:

Canonical link: https://commits.webkit.org/301987@main
